system/lib/wasmfs/memory_backend.h

```diff
@@ -90,6 +88,12 @@ class MemoryDirectory : public Directory {
   size_t getNumEntries() override { return entries.size(); }
   std::vector<Directory::Entry> getEntries() override;
 
+  std::string getName(std::shared_ptr<File> file) override;
+
+  // Since we internally track files with `File` objects, we don't need the
+  // dcache as well.
+  bool maintainsFileIdentity() override { return true; }
```

Comment:
Reading this code, I'm a little lost. Why does `MemoryDirectory` not just let the dcache do its thing? I don't see any comments explaining that and I can't figure it out myself...

The dcache maps names to Files, and so does this class, just it does it in a vector instead of a map..?

Reply:
Either way works. The dcache isn't necessary for `MemoryDirectory`, but it doesn't get in the way either. The benefit of disabling the dcache for `MemoryDirectory` in this PR is mostly just to prove that a backend that disables the dcache can work at all. It will be more concretely useful in a case-insensitivity backend and in the future once we get around to making the node backend more robust in the presence of case-insensitive underlying file systems.

Reply:
If either way works, can we greatly simplify `MemoryBackend` by removing it's code that duplicates the dcache functionality?

Reply:
Can you explain that part to me, then? That's what I was asking before. I'm missing something that is preventing me from understanding this code enough to review it. Specifically, why couldn't it be more minimal before this PR? In other words, if it is doing the same as the dcache, why couldn't we just use the dcache there?

Reply:
The dcache has to sit on top of some other source of truth. It doesn't have the ability to create files on its own, for instance, even though it can cache the existence of those files once they have been created. That's why the memory directory had to implement file creation. The reason the memory directory had to keep a list of all the files it has created is because the dcache never knows if it has a complete picture of the underlying directory, so it always has for forward `getEntries` to the backend. I think that pretty much covers all the things the memory directory previously implemented. This PR additionally makes it implement `getChild` and `getName`, which were previously handled entirely by the dcache.

Reply:
So, the `dcache` only caches files but not owns them while `MemoryBackend::entries` owns files it has, right?
I was a bit confused at first when I saw that, so decided  to ask here.

Reply:
Yeah, that's a good summary 👍 This stuff is definitely complicated...